test-data/unit/check-attr.test

```diff
@@ -0,0 +1,160 @@
+
+[case testAutoAttribAttrS]
+import attr
+import typing
+@attr.s(auto_attribs=True)
+class Auto:
+    normal: int
+    _private: int
+    def_arg: int = 18
+    _def_kwarg: int = attr.ib(validator=None, default=18)
+
+    UNTYPED_CLASS_VAR = 18
+    TYPED_CLASS_VAR: typing.ClassVar[int] = 18
+
+reveal_type(Auto)  # E: Revealed type is 'def (normal: builtins.int, private: builtins.int, def_arg: builtins.int =, def_kwarg: builtins.int =) -> __main__.Auto'
+Auto(1, 2)
+Auto(1, 2, 3)
+Auto(1, 2, 3, 4)
+Auto(1, 2, def_kwarg=5)
+Auto(1, 2, def_kwarg=5)
+Auto(normal=1, private=2, def_arg=3, def_kwarg=4)
+Auto()  # E: Too few arguments for "Auto"
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
+
+[case testSeriousNamesAttrS]
+import typing
+from attr import attrib, attrs
+@attrs(auto_attribs=True)
+class Auto:
+    normal: int
+    _private: int
+    def_arg: int = 18
+    _def_kwarg: int = attrib(validator=None, default=18)
+    UNTYPED_CLASS_VAR = 18
+    TYPED_CLASS_VAR: typing.ClassVar[int] = 18
+reveal_type(Auto)  # E: Revealed type is 'def (normal: builtins.int, private: builtins.int, def_arg: builtins.int =, def_kwarg: builtins.int =) -> __main__.Auto'
+Auto(1, 2)
+Auto(1, 2, 3)
+Auto(1, 2, 3, 4)
+Auto(1, 2, def_kwarg=5)
+Auto(1, 2, def_kwarg=5)
+Auto(normal=1, private=2, def_arg=3, def_kwarg=4)
+Auto()  # E: Too few arguments for "Auto"
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
+
+[case testNoInitAttrS]
+from attr import attrib, attrs
+@attrs(auto_attribs=True, init=False)
+class Auto:
+    normal: int
+    _private: int
+    def_arg: int = 18
+    _def_kwarg: int = attrib(validator=None, default=18)
+reveal_type(Auto)  # E: Revealed type is 'def () -> __main__.Auto'
+Auto()
+Auto() == Auto()
+Auto() >= Auto()
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
+
+[case testNoCmpAttrS]
+from attr import attrib, attrs
+@attrs(auto_attribs=True, cmp=False)
+class A:
+    b: int
+reveal_type(A)  # E: Revealed type is 'def (b: builtins.int) -> __main__.A'
+A(5)
+A(5) == A(5)
+A(5) >= A(5)  # E: Unsupported left operand type for >= ("A")
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
+
+
+[case testInheritanceAttrS]
+import attr
+@attr.s
+class A:
+    a: int = attr.ib()
+@attr.s
+class B:
+    b: str = attr.ib()
+@attr.s  # type: ignore  # Incompatible base classes in re cmp methods
+class C(B, A):
+    c: bool = attr.ib()
+reveal_type(C)  # E: Revealed type is 'def (a: builtins.int, b: builtins.str, c: builtins.bool) -> __main__.C'
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
+
+[case testNestedClassAttrS]
+import attr
+@attr.s
+class C:
+    y = attr.ib()
+    @attr.s
+    class D:
+        x: int = attr.ib()
+reveal_type(C)  # E: Revealed type is 'def (y: Any) -> __main__.C'
+reveal_type(C.D)  # E: Revealed type is 'def (x: builtins.int) -> __main__.C.D'
+[builtins fixtures/attr_builtins.pyi]
+[add-module fixtures/attr.pyi]
```

Comment:
Could you please add more tests? Check not only that definitions are processed correctly, but how they interact with various other concepts. For example, try to have good coverage at least for:
* type aliases (both inside and outside of a class)
* type variables
* decorated generic classes
* type inference
* forward references (both inside and outside of class)
* importing decorated classes
* instance and class methods in decorated types

Reply:
The only tests I couldn't write because I wasn't sure what you meant were "type inference"  But I think I covered everything else.  Let me know if you want more tests.